cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/ProgramScheduleHttpHandler.java

```diff
@@ -0,0 +1,813 @@
+      @QueryParam("trigger-type") String triggerType,
+      @QueryParam("schedule-status") String scheduleStatus) throws Exception {
+    ProgramType programType = ProgramType.valueOfCategoryName(type, BadRequestException::new);
+    if (programType.getSchedulableType() == null) {
+      throw new BadRequestException("Program type " + programType + " cannot have schedule");
+    }
+
+    ProgramId programId = new ApplicationId(namespace, application).program(programType, program);
+    doGetSchedules(responder, new NamespaceId(namespace).app(application), programId,
+        triggerType, scheduleStatus);
+  }
+
+  private void doGetSchedules(HttpResponder responder, ApplicationId applicationId,
+      @Nullable ProgramId programId, @Nullable String triggerTypeStr,
+      @Nullable String statusStr) throws Exception {
+    ApplicationSpecification appSpec = Optional.ofNullable(
+            store.getLatest(applicationId.getAppReference()))
+        .map(ApplicationMeta::getSpec)
+        .orElse(null);
+    if (appSpec == null) {
+      throw new NotFoundException(applicationId);
+    }
+    ProgramScheduleStatus status;
+    try {
+      status = statusStr == null ? null : ProgramScheduleStatus.valueOf(statusStr);
+    } catch (IllegalArgumentException e) {
+      throw new BadRequestException(
+          String.format("Invalid schedule status '%s'. Must be one of %s.",
+              statusStr, Joiner.on(',').join(ProgramScheduleStatus.values())), e);
+    }
+
+    ProtoTrigger.Type triggerType;
+    try {
+      triggerType =
+          triggerTypeStr == null ? null : ProtoTrigger.Type.valueOfCategoryName(triggerTypeStr);
+    } catch (IllegalArgumentException e) {
+      throw new BadRequestException(e.getMessage(), e);
+    }
+
+    Predicate<ProgramScheduleRecord> predicate = record -> true;
+    if (status != null) {
+      predicate = predicate.and(record -> record.getMeta().getStatus().equals(status));
+    }
+    if (triggerType != null) {
+      predicate = predicate.and(
+          record -> record.getSchedule().getTrigger().getType().equals(triggerType));
+    }
+
+    Collection<ProgramScheduleRecord> schedules;
+    if (programId != null) {
+      if (!appSpec.getProgramsByType(programId.getType().getApiProgramType())
+          .contains(programId.getProgram())) {
+        throw new NotFoundException(programId);
+      }
+      schedules = programScheduleService.list(programId, predicate);
+    } else {
+      schedules = programScheduleService.list(applicationId, predicate);
+    }
+
+    List<ScheduleDetail> details = schedules.stream()
+        .map(ProgramScheduleRecord::toScheduleDetail)
+        .collect(Collectors.toList());
+    responder.sendJson(HttpResponseStatus.OK,
+        ProgramHandlerUtil.toJson(details, Schedulers.SCHEDULE_DETAILS_TYPE));
+  }
+
+  /**
+   * Returns the previous runtime when the scheduled program ran.
+   */
+  @GET
+  @Path("/apps/{app-name}/{program-type}/{program-name}/previousruntime")
```

Comment:
`program-type` can be replaced by `schedules`.

Similar comment for all such request paths.